resource/environment/integration.js

```diff
@@ -241,6 +245,141 @@ function upgrade( slack, envelope ) {
 	} );
 }
 
+function getTemplate( token, catalogOwner, catalog, info ) {
+	function onRancherCompose( templateResult, rancherCompose ) {
+		templateResult["rancher-compose.yml"] = rancherCompose;
+		return templateResult;
+	}
+	function onDockerCompose( templateResult, response, dockerCompose ) {
+		templateResult["docker-compose.yml"] = dockerCompose;
+		for ( var i = 0; i < response.length; i++ ) {
+			if ( response[i].name === "rancher-compose.yml" ) {
+				return rp( response[i].download_url, {
+					qs: {
+						access_token: token
+					},
+					headers: {
+						"User-Agent": "cowpoke"
+					}
+				} ).then( onRancherCompose.bind( null, templateResult ) );
+			}
+		}
+	}
+	function onTemplate( v, response ) {
+		 var templateResult = {
+			version: v
+		};
+		var files = [];
+		for ( var i = 0; i < response.length; i++ ) {
+			if ( response[i].name === "docker-compose.yml" ) {
+				return rp( response[i].download_url, {
+					qs: {
+						access_token: token
+					},
+					headers: {
+						"User-Agent": "cowpoke"
+					}
+				} ).then( onDockerCompose.bind( null, templateResult, response ) );
+			}
+		}
+	}
+	function onResponse( response ) {
+		return when.all( _.map( response, function( dir ) {
+			if ( isnum( dir.name ) ) {
+				return rp( dir._links.self, {
+					qs: {
+						access_token: token
+					},
+					headers: {
+						"User-Agent": "cowpoke"
+					},
+					json: true
+				} ).then( onTemplate.bind( null, dir.name ) );
+			}
+		} ) ).then( function( templates ) {
+			return _.filter( templates, function( template ) {
+				if ( !template ) {
+					return false;
+				}
+				var dockerCompose = yaml.safeLoad( template["docker-compose.yml"] );
+				return util.getImageInfo( dockerCompose[info.docker.image + "-app"].image ).newImage === info.newImage;
+			} );
+		} ).then( function name( resultsAsArray ) {
+			return resultsAsArray[0];
+		} );
+	}
+
+	return rp( format( "https://api.github.com/repos/%s/%s/contents/templates/%s?ref=master", catalogOwner, catalog, info.branch ), {
+			qs: {
+				access_token: token
+			},
+			headers: {
+				"User-Agent": "cowpoke"
+			},
+			json: true
+		} ).then( onResponse ).catch( console.log );
+}
+
+function upgradeStack( slack, github, envelope ) {
```

Comment:
it's a little difficult to reason about the code b/c of the amount of nested functions. may want to consider modularize this into modules